### PR TITLE
EHN: ASC/DESC ordering in sort argsort, and argpartition.

### PR DESCRIPTION
**Pull request by rainwoodman**
(A stab at Issue #6217)

c_metadata is used to pass around pre-compiled
order_flags/offsets/descr. This should have made VOID_compare slightly
faster (avoiding calling python runtime) when called from within the
sorting routines.

Note that although c_metadata was claimed to be used for this purpose,
(see https://github.com/numpy/numpy/blob/master/numpy/core/include/numpy/ndarraytypes.h#L284)
I could not find an example of using c_metadata this way in numpy source tree.

The practical reason c_metadata is used is mainly to workaround the 'convention'
that the _compare functions take an array instance as the third argument.

Deeper inside, c_metadata is a workaround of two issues:

1) The offset and sub type-descrs are not exposed to NPY C-API directly.
   Hence, VOID_Compare had to walk the Python object hierarchy to retrieve
   such information. With this patch argsort/sort/argpartition
   will set up c_metadata to fill in this gap. But calling VOID_Compare from
   other places will still take the slower routine (PyDict/PyTuple)

2) Ordering of VOID type can not be uniquely defined by the ->compare functions
   of sub-descrs. The degeneracy in priority was 'broken(resolved)'
   by the _newnames trick in '_internal.py',
   but the degeneracy in the ranking of fields (ASC/DESC) could not be represented
   by any existing fields in a PyArray_Descr object. For example, for a compound structure
   dtype LastName, FirstName, each of the folowing defines a VOID_compare:

```
('LastName', ASC), ('FirstName', ASC)
('LastName', ASC), ('FirstName', DESC)
('LastName', DESC), ('FirstName', DESC)
('FirstName', ASC), ('LastName', ASC)
('FirstName', ASC), ('LastName', DESC)
('FirstName', DESC), ('LastName', DESC)
```

A further possible change after this patch is to release the GIL in argsort/sort for struct
dtypes, if it hasn't been released.

### CODE REVIEW

**Comment by rainwoodman** [comment]
BTW: I couldn't find where to update the documents of argsort / sort / argpartition.

**Comment by jaimefrio** [comment]
Nice, I wasn't aware of `NpyAuxData`. Some thoughts:

 * I think we should get the functionality right before we worry about performance. It seems to me that `SortOrderAuxData` is unnecessarily complicated, since all you really need is a tuple of  `(field_name, sort_direction)` pairs, no? I see too many `malloc` calls and too few tests to not be a little scared about the current implementation.
 * Linked to the above, do we really need to move the parsing of `order` into the C code? If it is only a performance thing, I'd rather see it work in Python first.
 * One shortcoming of the current `order` argument is that it cannot be used to leave some fields out of the comparison. I think your implementation changes this, by ignoring unspecified fields. While that change cannot be made, because it breaks backwards compatibility, it would be nice if we could set the direction to e,g, `0`, and have it mean "ignore this field".
 * The docstrings of those functions are set in `numpy/core/fromnumeric.py`, see [here](https://github.com/numpy/numpy/blob/maintenance/1.10.x/numpy/core/fromnumeric.py#L823) for `argsort`'s.
 * Rather than adding another argument, we should probably repurpose `order` for types with no fields, so that e.g. `order=-1` would mean "sort in decreasing order."

But very nice work! it may take some time and effort to get this fully figured out, but I think it will be well worth it.

**Comment by rainwoodman** [comment]
- I reduced the malloc calls to a single. What is the preferred way of allocating memory in numpy? This is still in GIL, so I am thinking maybe we shall go with PyMem_Malloc instead?

- I added a first stage parsing of the tuple in python (_internal.py:: _parse_order). It normalizes the order argument such that it is either -1 or a list of tuples. This hugely simplified the C counter-part (no need to raise exceptions).

- The constrain on order parameter is introduced in #5927. I have been using argsort for years and, based on what I read from the document (up to 1.9, http://docs.scipy.org/doc/numpy/reference/generated/numpy.sort.html), there was never a hint the unlisted fields are used in sorting. I always assumed the behaviour of argsort on unlisted fields is undefined. Because of this ambiguity, I would consider both behaviours conforming to the public document up to 1.9, and not a 'severe' backward incompatibility (if it is one at all)

I understand #5927 was to describe the exact behaviour of the code. But this previously undocumented behaviour does seem a bit convoluted. Also notice #5927 is less than 2 month old, and there probably is not even a public release since then.

- I updated docstrings to give examples how to use the extended list form of order argument.

- The post processing of -1 still needs to be done. I am not sure where is the shortest path. Do we really need to modify both sort and argsort, differently?

**Comment by rainwoodman** [comment]
Oh, I also added 0 for `ignoring the field`. But my point on the unlisted fields are the same.

**Comment by njsmith** [comment]
The use of `descr->c_metadata` is really really ugly :-(

Also, `c_metadata` is terrible to start with, and we should get rid of it indepedently of this (cf. the whole discussion about not inventing our own type system from scratch for dtypes)

Also, to the extent that `c_metadata` does exist and is used, then this code will be totally broken b/c it will overwrite whatever was supposed to be there. (I think ATM only datetime64-related stuff uses it.)

It's a clever hack but I think I am -1 on this implementation.

We need to rethink how sort works in general to make it a gufunc... don't think that the current code organization is sacred :-)

**Comment by ahaldane** [comment]
@njsmith I was worried about overwriting `c_metadata` too, but I think it's not a problem: This PR creates a new dummy dtype object and overwrites *that* object's `c_metadata`, preserving the old dtype's `c_metadata`. The user never gets exposed to a dtype with non-null `c_metadata`, it's only used internally for the duration of `array_sort`, and only for known structured types. 

Although in the long run there are no doubt better solutions, this seems like a pretty neat solution to me in the meantime. It will be nice to have the functionality and unit tests there when trying to improve dtypes/sorting.

File: numpy/core/src/multiarray/methods.c [hunk 1]
```diff
@@ -119,6 +120,62 @@ forward_ndarray_method(PyArrayObject *self, PyObject *args, PyObject *kwds,
...
+static PyArray_Descr * 
+parse_order(PyArray_Descr *descr, PyObject * order) 
+{
+    /* This function parses order argument and creates a new descr
+     * with c_metadata ready to be used by VOID_Compare */
+    npy_intp i;
+    PyArray_Descr *newd = NULL;
+    SortOrderAuxData *p = NULL;
+    npy_intp n_fields;
+    PyObject * _parse_order = NULL;
```
Comment (ahaldane) [inline numpy/core/src/multiarray/methods.c]: this should be static (so `npy_cache_import` only imports it the first time)

**Comment by ahaldane** [comment]
I think this is cool. 

I think @jaimefrio was suggesting simplifying the `SortOrderAuxData` structure so it only has two variables: `{ PyObject *field_name; int flag; }` and then looking up all the field info each time in `VOID_compare`. It would be slow, but avoids "dangerous" malloc/free manipulations. 

I would keep the code with all the mallocs around somewhere though, so we can add it back in later and benchmark. I thought it looked good, and I like the speed for structured array sorting.

**Comment by rainwoodman** [comment]
- On c_metadata:

@ahaldane is right the current PR does not break current use of c_metadata. I was lured into this from the comment in ndarraydtype.h claiming c_metadata should be used for these things. 

- One step further on c_metadata

Overall, if the major code structure is not to change, then it can be handy to allow attaching arbitrary metadata in the C-API to a type descr or an array, for an array seems to be the only context that is passed around in callback functions in numpy. 

This seems to be a widely used pattern in callback heavy systems. For example, GObject API has g_object_set_data and g_object_get_data, https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#g-object-get-data .

We can easily avoid overwriting, by promoting c_metadata to a link-list, and 
add two macros / functions for finding and removing 'my' metadata, something along the lines of `NPY_AUX_DATA_GET(metadata, clone_func)`. 

But I suspect this is becoming the type system that @njsmith mentioned people wanted to avoid.

- On 'malloc': 

the current PR only has one malloc, which is needed to allocate the `soad` in any case, no matter the descr is pre-calculated by parse_order or in VOID_compare. So, it has to be there. The leaks in 624f08c has been fixed. In other words, passing in descr from soad does not need more malloc, and we get some speed for free. (see below)

- Speed:

On my desktop, argsort by 26 columns on 1,000,000 'f8' rows, (normal distribution with random seed 12345), this PR takes 0.65 seconds. 4ea71a6 takes 0.97 seconds. About 30% decrease in time.

- order == -1

We can probably patch each sort function in npysort/ to call the 'reversed' version based on the flags of another c_metadata (sorry @njsmith), which could be dangerous. (Can I duplicate a static dtype?)

With a post processing approach, I am sensing quite a lot of new functions in arraytype.c.src . @jaimefrio Do you see the same?

**Comment by rainwoodman** [comment]
I think I've found an easy solution to order==-1 case.

**Comment by rainwoodman** [comment]
Is it still desired to add this feature before a dtype rewrite?

In the last a few commits 
- order == -1 is supported,
- documentation updated, hopefully more readable.
- ensuring a new view of array together with a new dtype.  The caller never sees the c_metadata which only lives during the sort. The new view may help eliminate possible race condition on the dtype
attribute of the original array, when (not now, but if ever) GIL is released.
- I implemented compatiblility with #5927 (default to ascending for unspecfied fields), but stashed it for some severe usability considerations. 

Consider the following example:
```
       dtype = np.dtype([
                ('id', 'i8'),
                ('First', 'S8'),
                ('Last', 'S8'),
                ])
        data = np.array(
            [ (0, 'Sam', 'Adams') ] * 100
        +   [ (1, 'Brian', 'Adams') ] * 100
        +   [ (2, 'Smith', 'Adams') ] * 100
            , dtype=dtype)
        data['id'] = np.arange(len(data)) % 100
```

With the default ascending behavior, the following line is *not* a stable sort on 'Last':
```
        a = data.argsort(kind='m', order=['Last'])
```

Rather one will need this
```
        a = data.argsort(kind='m', order=['Last', ('id', 0), ('First', 0)])
```

or
```
       a = data.argsort(kind='m', order=['Last'] + [(name, 0) for name in data.dtype.names if name != 'Last')] )
```
Therefore, the #5927 syntax requires the caller to do more work for a more common use case, encouraging unintuitive code. Is this really something we are after?

**Comment by rainwoodman** [comment]
Shall I spend more time to bring this up to date or is this patch dead because it touched internal parts of numpy that people dislike?

**Comment by eric-wieser** [comment]
Related to #14728

**Comment by rainwoodman** [comment]
c_metadata is still there. Resolving the conflicts is likely straight-forward.

**Pull request closed without merging**